File: artist_lead_scraper.js
const chromium = require('@sparticuz/chromium');
const puppeteer = require('puppeteer-core');
const fs = require('fs');
const path = require('path');

// Helper to pause
const sleep = ms => new Promise(res => setTimeout(res, ms));

// Helper to decode SoundCloud's gate.sc URLs
function decodeGateSc(url) {
  try {
    const u = new URL(url);
    if (u.hostname === 'gate.sc' && u.searchParams.has('url')) {
      return decodeURIComponent(u.searchParams.get('url'));
    }
  } catch (e) {}
  return url;
}

// Extract producer channel names from YouTube search
async function getTopProducersFromYouTube(searchTerm, numResults = 5) {
  const browser = await puppeteer.launch({
    args: chromium.args,
    defaultViewport: chromium.defaultViewport,
    executablePath: await chromium.executablePath(),
    headless: chromium.headless,
    ignoreHTTPSErrors: true,
  });
  
  const page = await browser.newPage();
  await page.goto(`https://www.youtube.com/results?search_query=${encodeURIComponent(searchTerm + ' type beat')}`);
  await page.waitForSelector('ytd-video-renderer', { timeout: 10000 });
  const producers = await page.evaluate((numResults) => {
    const results = [];
    const videos = Array.from(document.querySelectorAll('ytd-video-renderer'));
    for (let i = 0; i < Math.min(numResults, videos.length); i++) {
      const channel = videos[i].querySelector('ytd-channel-name a');
      if (channel) {
        let name = channel.textContent.trim();
        // Remove common suffixes
        name = name.replace(/\s*(type\s*)?beats?\s*$/i, '');
        results.push(name);
      }
    }
    return results;
  }, numResults);
  await browser.close();
  return producers;
}

// Scrape SoundCloud for artists using a producer's beats
async function getArtistsFromSoundCloud(producerName, maxArtists = 5) {
  const browser = await puppeteer.launch({
    args: chromium.args,
    defaultViewport: chromium.defaultViewport,
    executablePath: await chromium.executablePath(),
    headless: chromium.headless,
    ignoreHTTPSErrors: true,
  });
  
  const page = await browser.newPage();
  const searchUrl = `https://soundcloud.com/search?q=prod.%20${encodeURIComponent(producerName)}`;
  await page.goto(searchUrl, { waitUntil: 'domcontentloaded' });
  await sleep(4000);
  // Get artist profile links from search results
  const artistLinks = await page.evaluate(() => {
    const links = Array.from(document.querySelectorAll('a.soundTitle__username'));
    return links.map(a => a.href);
  });
  const uniqueLinks = [...new Set(artistLinks)].slice(0, maxArtists);
  const artists = [];
  for (const link of uniqueLinks) {
    const info = await scrapeArtistProfile(link, browser);
    if (info) artists.push(info);
    await sleep(2000);
  }
  await browser.close();
  return artists;
}

// Scrape a SoundCloud artist profile for contact info
async function scrapeArtistProfile(artistUrl, browser) {
  const page = await browser.newPage();
  await page.goto(artistUrl, { waitUntil: 'domcontentloaded' });
  await sleep(3000);
  let artistInfo = {
    url: artistUrl,
    name: '',
    email: '',
    instagram: '',
    twitter: '',
    youtube: '',
    website: '',
    bio: ''
  };
  try {
    // Name
    await page.waitForSelector('h2.profileHeaderInfo__userName', { timeout: 10000 });
    artistInfo.name = await page.$eval('h2.profileHeaderInfo__userName', el => el.textContent.trim());
    // Bio as HTML and text
    let bioHtml = '';
    try {
      bioHtml = await page.$eval('div.profileHeaderInfo__bio', el => el.innerHTML);
      artistInfo.bio = await page.$eval('div.profileHeaderInfo__bio', el => el.textContent.trim());
    } catch {}
    // Extract all emails from mailto links in bio HTML
    const emailsFromMailto = await page.$$eval('div.profileHeaderInfo__bio a[href^="mailto:"]', links =>
      links.map(a => a.href.replace(/^mailto:/, '').trim())
    );
    // Extract all emails from text (in case there are more)
    const emailsFromText = (artistInfo.bio.match(/[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g) || []).map(e => e.trim());
    // Combine and deduplicate emails
    const allEmails = Array.from(new Set([...emailsFromMailto, ...emailsFromText]));
    if (allEmails.length > 0) {
      artistInfo.email = allEmails.join(', ');
    }
    // Social links from buttons (e.g. IG, YT)
    const socialLinks = await page.$$eval('a.sc-social-logo-interactive', links =>
      links.map(a => ({
        href: a.href,
        label: a.textContent.trim().toLowerCase()
      }))
    );
    for (const { href, label } of socialLinks) {
      const realUrl = href.includes('gate.sc') ? (new URL(href)).searchParams.get('url') ? decodeURIComponent((new URL(href)).searchParams.get('url')) : href : href;
      if (realUrl.includes('instagram.com') && !artistInfo.instagram) artistInfo.instagram = realUrl;
      if (realUrl.includes('youtube.com') && !artistInfo.youtube) artistInfo.youtube = realUrl;
      if (realUrl.includes('twitter.com') && !artistInfo.twitter) artistInfo.twitter = realUrl;
      if (!artistInfo.website && label === 'website') artistInfo.website = realUrl;
    }
    // Extract Instagram from patterns like 'Insta: username', 'IG: username', 'Instagram: username'
    if (!artistInfo.instagram) {
      const instaPattern = /(?:Insta|IG|Instagram)[:\s]+([a-zA-Z0-9._-]+)/i;
      const instaMatch = artistInfo.bio.match(instaPattern);
      if (instaMatch) {
        artistInfo.instagram = `https://instagram.com/${instaMatch[1]}`;
      }
    }
    // Also keep @username pattern as fallback (but avoid emails)
    if (!artistInfo.instagram) {
      const atUserMatches = artistInfo.bio.match(/@[a-zA-Z0-9._-]+/g);
      if (atUserMatches && atUserMatches.length > 0) {
        // Filter out @ signs that are part of emails
        const igHandle = atUserMatches.find(u => !allEmails.some(e => e.includes(u)));
        if (igHandle) {
          artistInfo.instagram = `https://instagram.com/${igHandle.replace('@', '')}`;
        }
      }
    }
    // Twitter from bio if not found
    if (!artistInfo.twitter) {
      const twPatterns = [
        /twitter\.com\/([\w.-]+)/i,
        /@([\w.-]+)\s*(?:on\s*)?(?:tw|twitter)/i,
        /(?:tw|twitter):\s*@?([\w.-]+)/i,
        /(?:tw|twitter)\s*@?([\w.-]+)/i
      ];
      for (const pattern of twPatterns) {
        const match = artistInfo.bio.match(pattern);
        if (match) {
          artistInfo.twitter = `https://twitter.com/${match[1]}`;
          break;
        }
      }
    }
    // Website from bio if not found
    if (!artistInfo.website) {
      const webPatterns = [
        /(https?:\/\/[^\s]+)/i,
        /(?:website|site|web):\s*(https?:\/\/[^\s]+)/i,
        /(?:website|site|web)\s*(https?:\/\/[^\s]+)/i
      ];
      for (const pattern of webPatterns) {
        const match = artistInfo.bio.match(pattern);
        if (match) {
          artistInfo.website = match[1];
          break;
        }
      }
    }
  } catch (e) {
    console.error(`Error scraping ${artistUrl}:`, e.message);
    await page.close();
    return null;
  }
  await page.close();
  return artistInfo;
}

// Export the functions for use in server.js
module.exports = {
    getTopProducersFromYouTube,
    getArtistsFromSoundCloud
};

File: artist_lead_scraper.py
from youtubesearchpython import VideosSearch
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from webdriver_manager.chrome import ChromeDriverManager
from bs4 import BeautifulSoup
import time
import re
import json
from typing import List, Dict
from urllib.parse import unquote, urlparse, parse_qs

class ArtistLeadScraper:
    def __init__(self):
        self.setup_driver()
        
    def setup_driver(self):
        """Set up the Chrome WebDriver with appropriate options."""
        chrome_options = Options()
        chrome_options.add_argument("--headless")  # Run in headless mode
        chrome_options.add_argument("--no-sandbox")
        chrome_options.add_argument("--disable-dev-shm-usage")
        
        service = Service(ChromeDriverManager().install())
        self.driver = webdriver.Chrome(service=service, options=chrome_options)
        
    def search_youtube_producers(self, search_term: str, num_results: int = 3) -> List[str]:
        """Search YouTube for beat producers and extract their names from channel names."""
        videos_search = VideosSearch(f"{search_term} type beat", limit=num_results)
        results = videos_search.result()['result']
        
        producers = []
        for video in results:
            # Extract producer name from channel name
            channel_name = video['channel']['name']
            # Remove common suffixes like "type beats" or "beats"
            producer_name = re.sub(r'\s*(type\s*)?beats?\s*$', '', channel_name, flags=re.IGNORECASE)
            producers.append(producer_name.strip())
        
        return producers
    
    def search_soundcloud_artists(self, producer_name: str) -> List[Dict]:
        """Search SoundCloud for artists using beats from the producer."""
        search_url = f"https://soundcloud.com/search?q=prod.%20{producer_name}"
        self.driver.get(search_url)
        
        # Wait for results to load
        time.sleep(5)  # Basic wait, could be improved with explicit waits
        
        # Get the page source and parse with BeautifulSoup
        soup = BeautifulSoup(self.driver.page_source, 'html.parser')
        
        artists = []
        # Find artist links in the search results
        artist_links = soup.find_all('a', {'class': 'soundTitle__username'})
        
        for link in artist_links:
            artist_url = f"https://soundcloud.com{link['href']}"
            artist_info = self.scrape_artist_info(artist_url)
            if artist_info:
                artists.append(artist_info)
        
        return artists
    
    def scrape_artist_info(self, artist_url: str) -> Dict:
        """Scrape contact information from an artist's SoundCloud profile."""
        self.driver.get(artist_url)
        time.sleep(3)  # Wait for profile to load
        
        artist_info = {
            'url': artist_url,
            'name': '',
            'email': '',
            'instagram': '',
            'twitter': '',
            'website': '',
            'youtube': '',
            'bio': ''  # Store the full bio for reference
        }
        
        try:
            # Get artist name
            name_element = WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located((By.CSS_SELECTOR, 'h1.profileHeaderInfo__title'))
            )
            artist_info['name'] = name_element.text
            
            # Get description which might contain contact info
            description = self.driver.find_element(By.CSS_SELECTOR, 'div.profileHeaderInfo__bio').text
            artist_info['bio'] = description
            
            # Extract email
            email_match = re.search(r'[\w\.-]+@[\w\.-]+\.\w+', description)
            if email_match:
                artist_info['email'] = email_match.group(0)
            
            # Extract social media links from social buttons
            social_links = self.driver.find_elements(By.CSS_SELECTOR, 'a.sc-social-logo-interactive')
            for link in social_links:
                href = link.get_attribute('href')
                label = link.text.strip().lower()
                # Decode gate.sc redirect if present
                if 'gate.sc' in href:
                    parsed = urlparse(href)
                    qs = parse_qs(parsed.query)
                    if 'url' in qs:
                        real_url = unquote(qs['url'][0])
                    else:
                        real_url = href
                else:
                    real_url = href
                if 'instagram.com' in real_url and not artist_info['instagram']:
                    artist_info['instagram'] = real_url
                if 'youtube.com' in real_url and not artist_info['youtube']:
                    artist_info['youtube'] = real_url
            
            # Extract social media links and mentions from bio (as before)
            instagram_patterns = [
                r'instagram\.com/([\w\.-]+)',
                r'@([\w\.-]+)\s*(?:on\s*)?(?:ig|insta|instagram)',
                r'(?:ig|insta|instagram):\s*@?([\w\.-]+)',
                r'(?:ig|insta|instagram)\s*@?([\w\.-]+)'
            ]
            for pattern in instagram_patterns:
                if artist_info['instagram']:
                    break
                instagram_match = re.search(pattern, description, re.IGNORECASE)
                if instagram_match:
                    username = instagram_match.group(1)
                    artist_info['instagram'] = f"https://instagram.com/{username}"
                    break
            twitter_patterns = [
                r'twitter\.com/([\w\.-]+)',
                r'@([\w\.-]+)\s*(?:on\s*)?(?:tw|twitter)',
                r'(?:tw|twitter):\s*@?([\w\.-]+)',
                r'(?:tw|twitter)\s*@?([\w\.-]+)'
            ]
            for pattern in twitter_patterns:
                if artist_info['twitter']:
                    break
                twitter_match = re.search(pattern, description, re.IGNORECASE)
                if twitter_match:
                    username = twitter_match.group(1)
                    artist_info['twitter'] = f"https://twitter.com/{username}"
                    break
            website_patterns = [
                r'https?://[^\s]+',
                r'(?:website|site|web):\s*(https?://[^\s]+)',
                r'(?:website|site|web)\s*(https?://[^\s]+)'
            ]
            for pattern in website_patterns:
                if artist_info['website']:
                    break
                website_match = re.search(pattern, description, re.IGNORECASE)
                if website_match:
                    artist_info['website'] = website_match.group(0)
                    break
        except Exception as e:
            print(f"Error scraping artist info: {str(e)}")
            return None
        return artist_info
    
    def close(self):
        """Close the WebDriver."""
        self.driver.quit()

def main():
    scraper = ArtistLeadScraper()
    try:
        # Search for Drake type beat producers
        producers = scraper.search_youtube_producers("Drake")
        print(f"Found producers: {producers}")
        
        all_artists = []
        for producer in producers:
            print(f"\nSearching for artists using beats from {producer}...")
            artists = scraper.search_soundcloud_artists(producer)
            all_artists.extend(artists)
        
        # Save results to a JSON file
        with open('artist_leads.json', 'w') as f:
            json.dump(all_artists, f, indent=2)
            
        print(f"\nFound {len(all_artists)} artists. Results saved to artist_leads.json")
        
    finally:
        scraper.close()

if __name__ == "__main__":
    main() 